src/librustc_mir/interpret/eval_context.rs

```diff
@@ -47,6 +47,22 @@ pub struct InterpretCx<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'a, 'mir, 'tcx>> {
         (Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>),
         Pointer<M::PointerTag>
     >,
+
+    /// Information about any in-progress panic unwinding
+    pub unwinding: UnwindingState
+}
+
+/// Keeps track of relevant unwinding information,
+/// across calls to pop_stack_frame
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum UnwindingState {
+    /// We are not unwinding. This is always the case in CTFE mode
+    NotUnwinding,
+    /// We are currently unwinding from a panic.
+    /// 'in_cleanup' indicates whether or not we are currently running the
+    /// 'unwind' block assocaited with a terminator. When 'in_cleanup' is true,
+    /// execution proceeds normally until we hit a 'Resume' terminator
+    Unwinding { in_cleanup: bool }
```

Comment:
Why is this information relevant? The comment should explain that. Isn't hitting `Resume` enough to know that we are unwinding?

Reply:
I wanted to ensure that we explicitly track whether or not we're currently unwinding. This allows us to detect double panics independent of libstd, and to detect any weird MIR (e.g. an unwind block that doesn't eventually lead to `Resume`).

Reply:
What is wrong with such an `Unwind` block? That's effectively catching a panic, right?

I'd say detecting double-panics independently of libstd is explicitly a non-goal. At least if it requires so much extra state and complexity as the tracking you are doing here.